src/OmniSharp.Script/ScriptProjectSystem.cs

```diff
@@ -128,37 +135,69 @@ public void Initalize(IConfiguration configuration)
             foreach (var inheritedCompileLib in inheritedCompileLibraries.SelectMany(x => x.ResolveReferencePaths()))
             {
                 _logger.LogDebug("Adding implicit reference: " + inheritedCompileLib);
-                AddMetadataReference(commonReferences, inheritedCompileLib);
+                AddMetadataReference(_commonReferences, inheritedCompileLib);
             }
 
             // Each .CSX file becomes an entry point for it's own project
             // Every #loaded file will be part of the project too
             foreach (var csxPath in allCsxFiles)
             {
-                try
-                {
-                    var csxFileName = Path.GetFileName(csxPath);
-                    var project = scriptHelper.CreateProject(csxFileName, commonReferences);
+                AddToWorkspace(csxPath);
+            }
 
-                    if (enableScriptNuGetReferences)
-                    {
-                        var scriptMap = compilationDependencies.ToDictionary(rdt => rdt.Name, rdt => rdt.Scripts);
-                        var options = project.CompilationOptions.WithSourceReferenceResolver(
-                            new NuGetSourceReferenceResolver(ScriptSourceResolver.Default,
-                                scriptMap));
-                        project = project.WithCompilationOptions(options);
-                    }
+            // Watch CSX files in order to add/remove them in workspace
+            _fileSystemWatcher.Watch(CsxExtension, OnCsxFileChanged);
+        }
 
-                    // add CSX project to workspace
-                    _workspace.AddProject(project);
-                    _workspace.AddDocument(project.Id, csxPath, SourceCodeKind.Script);
-                    _projects[csxPath] = project;
-                    _logger.LogInformation($"Added CSX project '{csxPath}' to the workspace.");
-                }
-                catch (Exception ex)
+        private void OnCsxFileChanged(string filePath, FileChangeType changeType)
+        {
+            if (changeType == FileChangeType.Unspecified && !File.Exists(filePath) ||
+                changeType == FileChangeType.Delete)
+            {
+                RemoveFromWorkspace(filePath);
+            }
+
+            if (changeType == FileChangeType.Unspecified && File.Exists(filePath) ||
+                changeType == FileChangeType.Create)
+            {
+                AddToWorkspace(filePath);
```

Comment:
Here we'll want to run `TryGetCompilationDependencies` again, because that would resolve `#r nuget:` and `#load nuget:` if these are used inside of the file. However, there is no API at the moment to do it other than for a current folder so this needs to go on a TODO list for later. I plan to adjust some APIs to facilitate that soon.

Reply:
Ah, I see. Dependencies are not added automatically by the NuGet Resolvers (`NuGetSourceReferenceResolver` and `NuGetMetadataReferenceResolver`)?

Reply:
That is the long term goal, but at the moment they can't due to a current limitation in Roslyn which prevents a reference from resolving into more than a single `MetadataReference`.

Reply:
Ok, in that case, will it add any value to run `TryGetCompilationDependencies` again here? Since a recently added file will most of the time be blank... Unless, it’s a rename, move or copy from within VSCode?

Reply:
I think for now let's just leave it as is - and when we upgrade to "self contained" resolvers the problem will fix itself. If that doesn't come soon enough, we'll do something smarter here - what you have in this PR is a big step forward anyway 👍

Reply:
Not to mention for the primary scenario we want to support - CSI - it will already work as expected.